Include examples from various sources

EmilyOng opened the pull request:
I have included several examples from: (more to come.)

## Examples

**[Prusti Closure Examples](https://github.com/FabianWolff/closure-examples)**

<details>
<summary>Summary</summary>

| Prusti | Heifer |
|--------|--------|
| [all](https://github.com/FabianWolff/closure-examples/blob/master/all.rs) | [src/prusti/all.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/all.ml) |
| [any](https://github.com/FabianWolff/closure-examples/blob/master/any.rs) | [src/prusti/any.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/any.ml) | 
| [blameassgn](https://github.com/FabianWolff/closure-examples/blob/master/blameassgn.rs) | [src/prusti/blameassgn.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/blameassgn.ml) | 
| [cl_returned](https://github.com/FabianWolff/closure-examples/blob/master/cl_returned.rs) | [src/prusti/cl_returned.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/cl_returned.ml) |
| [counter](https://github.com/FabianWolff/closure-examples/blob/master/counter.rs) | [src/prusti/counter.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/counter.ml) |
| [delegation](https://github.com/FabianWolff/closure-examples/blob/master/delegation.rs) | [src/prusti/delegation.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/delegation.ml) |
| [filter](https://github.com/FabianWolff/closure-examples/blob/master/filter.rs) | [src/examples/later/filter.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/examples/later/filter.ml) |
| [fold](https://github.com/FabianWolff/closure-examples/blob/master/fold.rs) | [src/examples/later/fold.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/examples/later/fold.ml) |
| [for_each](https://github.com/FabianWolff/closure-examples/blob/master/for_each.rs) | [src/examples/later/foreach.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/examples/later/foreach.ml) |
| [map](https://github.com/FabianWolff/closure-examples/blob/master/map_vec.rs) | [src/examples/map.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/examples/map.ml) |
| [repeat_with_n](https://github.com/FabianWolff/closure-examples/blob/master/repeat_with_n.rs) | [src/prusti/repeat_with_n.ml](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/prusti/repeat_with_n.ml) |
</details>

**CS1101S Programs**

<details>
<summary>Summary</summary>

- [Scoping section in Reading Assessment 1 20/21 S1](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/cs1/scoping.ml)
- [Correctness section in Reading Assessment 1 18/19 S1](https://github.com/EmilyOng/AlgebraicEffect/blob/EmilyOng/examples/src/cs1/correctness.ml)
</details>

**[Program Proofs Book](https://program-proofs.com/code.html)**

<details>
<summary>Summary</summary>

Selection of examples and exercises in
- [Chapter 6](https://github.com/EmilyOng/AlgebraicEffect/tree/EmilyOng/examples/src/program-proofs/chapter06)
- [Chapter 8](https://github.com/EmilyOng/AlgebraicEffect/tree/EmilyOng/examples/src/program-proofs/chapter08)
</details>

**Papers**

<details>
<summary>Summary</summary>

- [Modular Specification and Verification of Closures in Rust](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/closure.ml#L34-L77)
</details>

**Misc.**

<details>
<summary>Summary</summary>

- [min_max_plus](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/closure.ml#L80-L86) from https://ilyasergey.net/CS6217/_static/slides/04-FunLog.pdf
- [interleave](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/list.ml#L14-L19) from https://cs3110.github.io/textbook/chapters/ds/exercises.html
- [subsequence](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/list.ml#L2-L7)
- [replace](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/list.ml#L9-L12)
- [find_index](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/list.ml#L25-L28) from https://v2.ocaml.org/api/List.html#VALfind_index
- [exists](https://github.com/EmilyOng/AlgebraicEffect/blob/49364407136c7fea04a2e5387c8aee69f28837a2/src/examples/list.ml#L30-L33) from https://v2.ocaml.org/api/List.html#VALexists
</details>

## Tests

Tested with `TEST=1 dune test`.

dariusf commented:
Thanks so much. Before merging, could you just check that everything in `src/examples` is working? That directory is [published](https://github.com/songyahui/AlgebraicEffect/blob/StagedSL/web/examples.sh) to the web demo, so it'd be good to keep it tidy.

It's fine for other directories to be WIP, though if they're done, feel free to add them to [run.t](https://github.com/songyahui/AlgebraicEffect/blob/StagedSL/src/programs.t/run.t) or the web demo.

To check the latter you can run tests using `dune test` (2 mins) and `PROVER=WHY3 dune test` (about 12 mins).

dariusf commented:
I fixed the substitution issue from earlier and made the tests more reliable. Also added some notes on configuring Why3 and some experimental tracing features, which help a lot when diving into the system.

EmilyOng commented on src/prusti/counter.ml:
```diff
@@ -0,0 +1,26 @@
+(*@ ens T @*)
+= let x = ref 0 in
+  let inc ()
+  = let r = !x in x := !x + 1; r in
```
I removed the specification for `inc` here because the [test report](https://github.com/EmilyOng/AlgebraicEffect/blob/21e63497cfe011f28979f37c7d62e6df277906b0/parsing/hiplib.ml#L754-L763) expects that functions with a `_false` suffix only sees false results.

EmilyOng commented on src/prusti/all.ml:
```diff
@@ -16,9 +15,36 @@ let rec all xs pred
 (* Unlike pure length, this is not provable because p on the left may have effects *)
 
-let test1 xs
+(* let test1 xs (* FIXME *)
```
Previously, I wrongly thought that these tests pass if we require `xs=[1;2;3;4]` instead of `xs->[1;2;3;4]` (https://github.com/EmilyOng/AlgebraicEffect/issues/19)

EmilyOng commented on src/prusti/all.ml:
```diff
@@ -16,9 +15,36 @@ let rec all xs pred
+(* let has_property p xs = all p xs *)
+
+let test_all_positive n
+(*@ ex r ys; all_positive(ys, r); req res=ys; ens r=true/\res=ys @*)
```
I added `req res=ys` because:

<details><summary>why3 file</summary>

```why
module M
  use int.Int
  use bool.Bool
  use list.List
  use all.Extras
  
  goal goal1:
    forall n : Int.int, res : List.list Int.int, v141 : List.list Int.int, v159 : 
          Bool.bool, v160 : Bool.bool, v161 : Int.int, v162 : List.list
                                                                Int.int, v184 : 
          Bool.bool, v194 : Bool.bool.
      ((((((v141 = v162) /\ (v194 = v159)) /\ (v160 = (Int.( > ) v161 0)))
           /\
           (v194 = v184)
           /\
           ((Int.( <= ) n 0) = false))
          /\
          (v184 = true)
          /\
          (res = (List.Cons 1 v141)))
         /\
         (((v141 = v162) /\ (v194 = v159)) /\ (v160 = (Int.( > ) v161 0)))
         /\
         (v194 = v184)
         /\
         ((Int.( <= ) n 0) = false))
        ->
        (exists r : Int.int.
           ((Bool.andb v160 v159) = true) /\ (res = (List.Cons v161 v162)))
end
```
</details> 

`v161 ` is universally quantified and the following counterexample by z3 shows the goal is false.

<details><summary>Counterexample</summary>

```
n : int = 1;
v159 : bool = true
v160 : bool = true
v161 : int = 17870
v184 :  bool = true
```
</details>

dariusf replied:
That's a creative solution. I guess the intent is one of these:

```ml
(*@ ex r; all_positive(res, r); ens r=true @*)

(*@ all_positive(res, true) @*)
```

Unfortunately both don't work at the moment because of how we handle entailments stage by stage. I'm thinking about how to support it.

This works though, since all_positive doesn't have higher-order parameters and can be pure:

```ml
(*@ ens all_positive(res)=true @*)
```

I pushed here to show how to make the function available to Why3.

EmilyOng commented:
> I fixed the substitution issue from earlier and made the tests more reliable. Also added some notes on configuring Why3 and some experimental tracing features, which help a lot when diving into the system.

Thanks! I have checked the programs in `examples/` are working using `(PROVER=WHY3) DEBUG=0 dune exec parsing/hip.exe src/examples/*.ml`.

dariusf commented:
Looks good! Feel free to merge whenever you're ready.
